Open tree branches when scrolling

coolapso opened the pull request:
<!-- If this is your first pull request for Fyne please read the contributor docs at:
https://github.com/fyne-io/fyne/wiki/Contributing.
Be sure that your work is based off `develop` branch. --> 

### Description:
<!-- A summary of the change included and which issue it addresses.
Please include any relevant motivation and background. -->

Fixes the following code note: 

```
// TODO scrolling to a node should open all parents if they aren't already
```

This one was quite a bit out of my depth, and tried my best to find a somewhat clean way of doing it, therefore I don't have any smart reasoning as to why it was done like this other than a "this is how I managed to get it done" 

let me know if you want any changes. 



https://github.com/user-attachments/assets/d29cef66-b5ec-4520-bbbb-f0391578db44



### Checklist:
<!-- Please tick these as appropriate using [x] -->

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:
<!-- Please delete these if not required for this PR -->

- [ ] Public APIs match existing style and have Since: line.
- [ ] Any breaking changes have a deprecation path or have been discussed.
- [ ] Check for binary size increases when importing new modules.

coolapso commented:
> Thanks for looking at this.
> 
> I think the expanding could be more efficient - instead of re-constructing the entire tree parent items you should be able to record the ids that lead to a child - then ensure that each of those ids are expanded. (though if you're using `walk` then `onNode` may need to be called after the child items in a branch scenario.
> 
> I would also factor this out into a separate method (maybe `expandParents`) so it's more readable and has a unit test.

:thinking:  I kinda tried that first time but miserably failed. will give it another shot and add a `expandParents(...)`

andydotxyz commented on widget/tree.go:
```diff
@@ -231,7 +231,41 @@ func (t *Tree) OpenBranch(uid TreeNodeID) {
-// Resize sets a new size for a widget.
+// Opens the branches leading to a node
+//
+// Since: 2.8
```
There is no since required for private methods.

coolapso replied:
removed

andydotxyz commented on widget/tree.go:
```diff
@@ -231,7 +231,41 @@ func (t *Tree) OpenBranch(uid TreeNodeID) {
+	}
+
+	t.ensureOpenMap()
+	var opened []TreeNodeID
```
(just a bool instead to indicate if refresh is required)

coolapso replied:
Think I understood what you meant, let me know if I got it right

andydotxyz reviewed (changes requested):
Thanks, this is looking good - a few cleanups and I think that the path slice should be ordered for avoidance of confusion

coolapso commented on widget/tree.go:
```diff
@@ -231,7 +231,41 @@ func (t *Tree) OpenBranch(uid TreeNodeID) {
+		return
+	}
+
+	if len(parents) == 0 {
```
done! :)

coolapso commented:
well copilot review were some good catches, but I am sorry for having it jumping here! Apparently I've been a good opensource citizen and GitHub bumped me to the pro version?? :shrug:  ... I've disabled it now.

andydotxyz commented on widget/tree.go:
```diff
@@ -231,7 +231,34 @@ func (t *Tree) OpenBranch(uid TreeNodeID) {
 	t.Refresh()
 }
 
-// Resize sets a new size for a widget.
```
You seem to have removed the Resize documentation

coolapso commented:
@andydotxyz  addressed all the comments made so far, let me know if there is anything else you need me to do, would be cool to get this in so I can finish https://github.com/fyne-io/fyne/pull/5999 before the all the knowledge about what I've done here goes into cold storage :sweat_smile: 

thank you :)
